lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -1527,6 +1527,265 @@ bool OptimizeBatchNorm::run(Function *F, const CompilationContext &cctx) {
   return changed;
 }
 
+/// If \p node has uses, and all of them have one user node, return this user
+/// node. Otherwise, return nullptr.
+static Node *getOnlyUser(Node &node) {
+  if (!node.hasUsers()) {
+    // No users.
+    return nullptr;
+  }
+  Node *first = node.getUsers().front().getUser();
+  for (auto &U : node.getUsers()) {
+    if (U.getUser() != first) {
+      // Multiple users.
+      return nullptr;
+    }
+  }
+  // One user.
+  return first;
+}
+
+static Constant *getConstant(const NodeValue &NV) {
+  Node *N = NV.getNode();
+  if (isa<QuantizeNode>(N)) {
+    N = N->getNthInput(0);
+  }
+  return dyn_cast<Constant>(N);
+}
+
+/// Checks that \p constant sub-tensors along \p chId axis repeat.
+/// It also extracts the repeated dimension values into \p result.
+static bool isConstBroadcasted(std::vector<float> &result, const Constant &tile,
+                               int32_t chId) {
+  // TODO: This limitation can be lifted, but that is for simplicity.
+  if (tile.getType()->dims().size() != 4) {
+    return false;
+  }
+  // TODO: We can also support quantized constants if there is a need in it.
+  if (tile.getType()->getElementType() != ElemKind::FloatTy) {
+    return false;
+  }
+  auto handle = tile.getPayload().getHandle<float>();
+  glow::dim_t n, h, w, c;
+  for (c = 0; c < tile.getType()->dims()[chId]; c++) {
+    std::vector<glow::dim_t> dims = {c, 0, 0, 0};
+    std::rotate(dims.begin(), dims.begin() + dims.size() - chId, dims.end());
+    const float expected = handle.at(llvm::ArrayRef<glow::dim_t>(dims));
+    for (n = 0; n < tile.getType()->dims()[(chId + 1) % 4]; n++) {
+      for (h = 0; h < tile.getType()->dims()[(chId + 2) % 4]; h++) {
+        for (w = 0; w < tile.getType()->dims()[(chId + 3) % 4]; w++) {
+          std::vector<glow::dim_t> dimsE = {c, n, h, w};
+          std::rotate(dimsE.begin(), dimsE.begin() + dimsE.size() - chId,
+                      dimsE.end());
+          if (handle.at(llvm::ArrayRef<glow::dim_t>(dims)) != expected) {
+            return false;
+          }
+        }
+      }
+    }
+    result[c] = expected;
+  }
+  return true;
+}
+
+/// Collects the longest chain of arithmetic operations with constants starting
+/// from \p start. Updates \p scale and \p bias as it collects the operands.
+/// \p returns the last node in the chain.
+static Node *collectArithmeticChain(Function *F, Node *start, Constant &scale,
```

Comment:
`node` will always be a `ConvolutionNode` or `BatchNormalizationNode` here, right? Can we change the `start` argument of `collectArithmeticChain()` to be a NodeValue where we pass in `CN.getResult()` or `BN.getResult()`? We should try to use NodeValues as much as possible (including the return type here), which would allow us to avoid e.g. `getNthResult()` usage with raw indices.

Reply:
Yes, `ConvolutionNode `or `BatchNormalizationNode`. Fixed